File: rd-kt/rd-framework/src/jvmTest/kotlin/com/jetbrains/rd/framework/test/cross/CrossTestServerBase.kt
package com.jetbrains.rd.framework.test.cross

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.test.cross.util.portFile
import com.jetbrains.rd.framework.util.NetUtils
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.threading.SingleThreadScheduler

abstract class CrossTestServerBase : CrossTestBase() {
    fun server(lifetime: Lifetime, port: Int? = null): IProtocol {
        scheduler = SingleThreadScheduler(lifetime, "SingleThreadScheduler")
        return Protocol(Serializers(), Identities(IdKind.Server), scheduler,
                SocketWire.Server(lifetime, scheduler, port, "DemoServer"), lifetime)
    }

    init {
        protocol = server(socketLifetime, NetUtils.findFreePort(0))

        portFile.printWriter().use { out ->
            out.println((protocol.wire as SocketWire.Server).port)
        }
    }
}

File: rd-cpp/src/rd_gen_cpp/test/cross_test/base/CrossTestServerBase.cpp
#include "CrossTestServerBase.h"

#include "wire/SocketWire.h"

#include <cstdint>
#include <memory>

namespace rd
{
namespace cross
{
CrossTestServerBase::CrossTestServerBase() : CrossTestBase()
{
	auto ptr = std::make_shared<SocketWire::Server>(socket_lifetime, &scheduler, 0, "TestServer");
	wire = ptr;
	protocol = std::make_unique<Protocol>(Identities::SERVER, &scheduler, wire, socket_lifetime);

	uint16_t port = ptr->port;
	std::ofstream file(port_file);
	file << port;
}
}	 // namespace cross
}	 // namespace rd